doc: Add Thinger.io Cloud Integration (#952)
TheThingsIndustries/lorawan-stack-docs: doc/content/integrations/cloud-integrations/thinger-io/_index.md

---
title: "Thinger.io"
description: ""
weight: 
---

[Thinger.io](https://thinger.io/) is a cloud IoT Platform that provides every needed tool to prototype, scale and manage connected products in a very simple way. Our goal is to democratize the use of IoT making it accessible to the whole world, and streamlining the development of big IoT projects.

> From [Thinger.io](https://thinger.io/) we wanted to offer an improved integration to {{% tts %}} users by providing easy to configure tools for the storing, analyzing and showing devices data in a simple way. [The plugin](https://docs.thinger.io/plugins/the-things-stack) allows to retrieve {{% tts %}} webhook messages to enhance the integration with some interesting features such as:

Reviewer: All fine, but can you please rephrase the **we** and **our** in these sentences? It's a bit weird to use **we** for Thinger.io when it's a TTS/TTI webpage

- Uplink data payload processing
- Downlink data payload processing
- Automatic device and data buckets provisioning

<!--more-->

## Requirements

1. Have an active subscription of Thinger.io Private Cloud Solution. Check [this link](https://thinger.io/pricing) to create your own private instance within minutes.
2. Install [{{% tts %}} Plugin](https://docs.thinger.io/plugins/the-things-stack) in the instance.

## Plugin Configuration

This section describes the different interfaces that can be used to configure the [{{% tts %}} Plugin](https://docs.thinger.io/plugins/the-things-stack).

### Integrating {{% tts %}} Applications

> The first step to perform the integration, is to create a new plugin configuration. It is possible to create multiple configuration profiles with custom behavior for each application deployed in {{% tts %}}. To create a new application profile, just type the Application ID and press the green **Add Application** button. Note that the ID must be exactly the same identifier defined in {{% tts %}} application.

Reviewer: ```suggestion
The first step to perform the integration, is to create a new plugin configuration on Thinger.io Cloud. It is possible to create multiple configuration profiles with custom behavior for each application deployed in {{% tts %}}. To create a new application profile, just type the **Application Id** and press the green **Add Application** button. Note that this ID must exactly match the application ID defined in your {{% tts %}} application.
```

{{< figure src="thingerio_default_settings.png" alt="[Thinger.io](https://thinger.io/) {{% tts %}} Plugin default settings" >}}

> The `Application Id` dropdown allows to select and configure a particular application profile, but if the "default" profile is selected, the configuration will be applied to all the applications integrated with the plugin.

Reviewer: ```suggestion
The `Application Id` dropdown allows to select and configure a particular application profile, but if the `Default` profile is selected, the configuration will be applied to all the applications integrated with the plugin.
```

- **Base64 String**: With binary information that can be sent directly to {{% tts %}} network. It is required if your {{% tts %}} application is not defining a converter.
- **JSON Object**: If {{% tts %}} application provides a converter for your payloads, this method can return a JSON object that will be accesible in the converter method. In this case, creating a custom downlink method will be redundant, so create the funtion in {{% tts %}}, or in the plugin.

Example of a downlink method converting a JSON device configuration into base64 as required by {{% tts %}}:

```javascript
/* convert a JSON object with the device configuration in a base64 string expected by {{% tts %}} */
module.exports.downlink = function(payload){
    let bytes = [];
    bytes[0] = payload.enabled ? 1 : 0;
    bytes[1] = payload.frequency;
    bytes[2] = payload.threshold;
    return Buffer.from(bytes).toString('base64');
};
```

{{< note >}}
The downlink method should return a base64 string if {{% tts %}} application does not define a converter.
{{</ note >}}

{{< /tabs/tab >}}

{{< /tabs/container >}}

## {{% tts %}} Console Configuration

### Uplink Configuration

> The last tab of the plugin configuration interface is called `Webhook Settings`, it has been created to help the developers to complete the integration in {{% tts %}} Console, by providing all the information required to setup the webhook profile. 

Reviewer: ```suggestion
The last tab of the plugin configuration interface is called **Webhook Settings**, it has been created to help the developers to complete the integration in {{% tts %}} Console, by providing all the information required to setup the webhook profile. 
```

{{< figure src="thingerio_webhook_settings.png" alt="[Thinger.io](https://thinger.io/) {{% tts %}} Plugin webhook settings tab" >}}

{{< note >}}
The REST API does not define the application ID, this parameter will be checked by the plugin software to manage the payload according to the configuration.
{{</ note >}}

To create a new webhook integration follow the next steps in {{% tts %}} web console:

1. Select the Application to be integrated.

> 2. In the main menu open the `Integrations` section and click the `Webhooks` option. The webhooks list will be shown.

Reviewer: ```suggestion
2. In the main menu open the **Integrations** section and click the **Webhooks** option. The webhooks list will be shown.
```

{{< figure src="thingerio_tts_sidebar.png" alt="{{% tts %}} sidebar" >}}

> 3. Clicking the **+Add webhook** blue button in the right top corner of the interface allows choosing between different webhooks integration templates. Select [Thinger.io](https://thinger.io/) template. Then, configure the webhook only requires filling the form with the information provided by [Thinger.io](https://thinger.io/) `Webhook Settings` tab and selecting JSON webhook format.

Reviewer: ```suggestion
3. Clicking the **+Add webhook** blue button in the right top corner of the interface allows choosing between different webhooks integration templates. Select [Thinger.io](https://thinger.io/) template. Then, configure the webhook only requires filling the form with the information provided by [Thinger.io](https://thinger.io/) **Webhook Settings** tab and selecting JSON webhook format.
```

{{< figure src="thingerio_tts_webhook.png" alt="[Thinger.io](https://thinger.io/) {{% tts %}} webhook settings" >}}

{{< note >}}

> The Authorization header must be set up using the access token including "Bearer" command

Reviewer: ```suggestion
The **Authorization Header** must be set up using the access token including `Bearer` command.
```

Reviewer: You can best spot the difference of using **this** and `this` on this example. For buttons, tabs, and field names in the Console we use **this**, while for variable values just like this one we use `this`. This is why I added some of these suggestions in the rest of the doc.
